### DISCUSSION TITLE
Use old-school generators to get async-await like flow for sync and async versions

### DISCUSSION THREAD

**Discussion by jcbhmr** [iter-tools/iter-tools#492] [Ideas]
Disclaimer: I know this probably won't be actually implemented in this codebase without massive overhauls. I just thought it was a cool idea that I'd share.

This code is seriously tough to me to grok. Sure, I can _get there_, but it takes like a minute to figure out "oh yeah that's what those fancy $await() is for" and "oh yeah that's right, the 'async;' statement is a _directive_, not code"

https://github.com/iter-tools/iter-tools/blob/975af78685efb8fe5b7963eba4342be9c1b2f348/src/impls/%24includes/%24includes.js#L1-L22

I then though of "wait, what about old-time async-await tricks where we used generator functions to do things almost just like this?"

https://medium.com/hackernoon/async-await-generators-promises-51f1a6ceede2

|||
|---|---|
|![](https://miro.medium.com/v2/resize:fit:720/format:webp/1*m6f3QQcaIkWN28CF5tU1gw.png)|![](https://miro.medium.com/v2/resize:fit:720/format:webp/1*Kyl2NKektJH_XsyjciS1BA.png)|

So I did something similar to that! I got it working with the simple case of an `includes()` function which only needs to return one thing: a boolean. If I were to expand this to include things like `map()` or `filter()`, I'd need to add in the capability to return an in-progress iterator that was wrapped with this maybe-await-maybe-sync generator thing. Check the comments! I may have an update in the future!

Here's a demo of includes() working: https://playcode.io/1221121

```js
// This will become a sync AND an async function!
function* includes(iterator, needle, comparator = (a, b) => a === b) {
  while (true) {
    // This 'yield' acts like an await THAT CAN BE SYNC
    const { value, done } = yield iterator.next();
    if (done) {
      return false
    }
    if (comparator(needle, value)) {
      return true;
    }
  }
}

function makeSync(g) {
  // This will be called like 'includes([1, 2, 3], 2)'
  return (iterable, ...args) => {
    const iterator = iterable[Symbol.iterator]();
    const state = g(iterator, ...args);
    let previousValue;
    while (true) {
      const { value, done } = state.next(previousValue);
      if (done) {
        return value;
      }
      previousValue = value;
    }
  }
}
function makeAsync(g) {
  // This will be called like 'includes(response.body, 2)'
  return async (iterable, ...args) => {
    const iterator = iterable[Symbol.asyncIterator]();
    const state = g(iterator, ...args);
    let previousValue;
    while (true) {
      // Our state g() function is SYNC. If it returns us a promise, we wait for it ourselves
      // so that the g() function is SYNC, we are the ones who are async.
      const { value, done } = state.next(previousValue);
      if (done) {
        return value;
      }
      previousValue = await value;
    }
  }
}

function* s() {
  yield "hello"
  yield " "
  yield "world"
}
async function* sAsync() {
  yield "hello"
  yield " "
  yield "world"
}

const includesSync = makeSync(includes)
const includesAsync = makeAsync(includes)

async function main() {
  console.log("sync A", includesSync(s(), "hello"))
  console.log("sync B", includesSync(s(), "56"))
  console.log("async A", await includesAsync(sAsync(), "hello"))
  console.log("async B", await includesAsync(sAsync(), "56"))
}
main()
```

```
sync A
true
sync B
false
async A
true
async B
false
```

**Comment by conartist6** [comment]
A lot of my other work like the regex engine and syntax-directed parser do use this approach. I know it works and I know how to implement it! And furthermore, I've actually investigated this quite a lot such that I already know how to do the transform wholesale on the library.

But! I'm never going to replace the sync variants, because my goal is to make the library better for users not better for maintainers. The sync variants are necessary because they are fast and simple. `function *map(fn, it) { for (const val of it) yield fn(it)` is basically a one-liner, so to be competitive I can't offer an experience substantially worse than writing the one-liner yourself or else nobody would use my library.

If I have to keep the plain sync variants, then it probably makes sense to keep the plain async variants, in part because their implementation is so obviously related to the implementation of the plain sync variants.

But but but! I actually think it's super interesting to define a new iteration protocol which is capable of returning items either synchronously OR asynchronously, depending on whether the items are available synchronously or not. You can read about that over here: https://es.discourse.group/t/take-2-generator-prototype-symbol-mixediterator/1308

**Reply by jcbhmr** [reply]
Wouldn't something like (below) mean that you a) keep all the ezpz imports (of course you'd need an index.js barrel import) while also helping regular JS devs grok it?

```js
// src/methods/includes.js
import { makeSync, makeAsync } from "..."
function* includesUnderlying(iterator, needle, comparator = (a, b) => a === b) {
  for (let value, done; !done; { value, done } = yield iterator.next()) {
    if (comparator(needle, value)) {
      return true;
    }
  }
  return false;
}
export const includes = makeSync(includesUnderlying)
export const asyncIncludes = makeAsync(includesUnderlying)
```

I can understand that UX takes priority over DX. That's valid. What I'm probably misunderstanding is this paragraph:

> But! I'm never going to replace the sync variants, because my goal is to make the library better for users not better for maintainers. The sync variants are necessary because they are fast and simple. function *map(fn, it) { for (const val of it) yield fn(it) is basically a one-liner, so to be competitive I can't offer an experience substantially worse than writing the one-liner yourself or else nobody would use my library.

I assume that this means you want to avoid getting into wrapper madness because it lowers performance? Valid. If you're talking about the `makeSync(includes)` thing, that can be pulled out of main() into the lib code.

**Reply by conartist6** [reply]
Lowering performance is one way of looking at it. Those yields back and forth do have cost, but what I care about more is that it adds cognitive overhead. For example it makes it hard to use a debugger if you don't understand exactly why `makeSync` is part of your code and what it is doing, which most people won't.

**Reply by conartist6** [reply]
Essentially it's an application of the Principle of Least Surprise

**Reply by jcbhmr** [reply]
> but what I care about more is that it adds cognitive overhead.

**In my opinion** this "principle of least surprise" would favor using native JS function wrappers instead of post-processing syntax? Maybe it's just because I've never seen code like this before lol.

I would think that the `makeSync()` or similar wrapper function is probably no different in "feel" or "brain pain" to the complicated currying magic that happens at the bottom of the file.

```js
 export const $includes = /*#__PURE__*/ $iterableCurry($__includes, { 
   minArgs: 1, 
   maxArgs: 2, 
   reduces: true, 
   validateArgs(args) { 
     if ($isSync && typeof args[0] === 'string') { 
       console.warn(`For string inputs use includesSeq instead of includes`); 
     } 
   }, 
 }); 
```

Probably _less brain pain_ since you can kinda get the gist that the magic generator function is handling the yield-to-await conversion. Especially if you use something like a `yield possiblyAwait(iterator.next())` function call.

I guess what I'm pointing out is that _I personally_ am gawking at this with my mouth wide open wondering how this all works. If I'm doing that, there's probably others. This _could_ be a problem if you (or whoever is the primary maintainer? I assume it's you) want average-Joe developers (like me) to feel like they can contribute code/docs/input/issues/etc. to this project. If you're fine with it being a solo project, then go for it! Nobody can stop you!

Possible solutions to this are to a) add more documentation and reference it whenever something gets weird or b) try to make the code "speak for itself" more so that you don't have to write so many docs.

It's almost like there's a complete DSL lurking in each `$include.js` file with a custom naming scheme and duplicate named files all over the place! It looks pretty complicated. 😳

**Reply by conartist6** [reply]
The currying magic is complicated, but you'll notice that I provided an escape hatch with `__`. It was important to me that `iter-tools` define and export the simplest version of each function.

**Reply by conartist6** [reply]
I freely admit that the contribution process is a currently a mess, primarily because my repos do some things in ways that nobody else does them. My current work is on building the tools I want and need and generalizing them sufficiently well that they can be used widely.

**Reply by jcbhmr** [reply]
I guess this comes down to the accessibility of a DSL extension to JS vs plain JS with weird generator function madness. Whichever one you end up deciding (or sticking with) is a-OK! I just am expressing my confuzzlement and the initial "hey! I wonder why they don't do that? Is there a reason?" knee-jerk that I had.

I think a great place for a "what the heck how does the `await;` directive work?" would be in a "DEVELOPMENT.md" or similar or at the bottom of the readme. But that's optional. Thanks for helping me understand more about how and why the magic await stuff works!

**Reply by conartist6** [reply]
I guess you didn't find https://github.com/iter-tools/iter-tools/blob/trunk/CONTRIBUTING.md? It has a link to the macro definition which has detailed documentation of the DSL including ASCII art.

**Reply by jcbhmr** [reply]
> I guess you didn't find https://github.com/iter-tools/iter-tools/blob/trunk/CONTRIBUTING.md? It has a link to the macro definition which has detailed documentation of the DSL including ASCII art.

Nope 🤦‍♂️ completely missed that! My bad sorry! 😬

**Comment by jcbhmr** [comment]
I actually got continuous generators working!

https://codesandbox.io/s/epic-dream-m9y7os?file=/script.js

Yes the code is really dirty and filled with single-character variable names. It's horrible. I know. 🤷‍♂️ But it is just the first draft, so that's OK! 😉

```js
// Mark a yield-ed value as needing "await" handled
const w = Symbol("maybeWait");
function maybeWait(o) {
  return { [w]: o };
}

// This will become a sync AND an async function!
function* map(f) {
  for (
    var { value, done } = yield maybeWait(this.next());
    !done;
    { value, done } = yield maybeWait(this.next())
  ) {
    yield f(value);
  }
  return value;
}

function makeAsyncContinuous(g) {
  async function* f(...args) {
    const m = g.call(this, ...args);
    for (
      var { value, done } = m.next();
      !done;
      { value, done } = m.next(value)
    ) {
      if (typeof value === "object" && value && w in value) {
        value = await value[w];
      } else {
        yield value;
      }
    }
    return value;
  }
  Object.defineProperty(f, "name", { value: g.name });
  Object.defineProperty(f, "length", { value: g.length });
  return f;
}

const mapAsync = makeAsyncContinuous(map);
```

![image](https://user-images.githubusercontent.com/61068799/220768179-19a44e61-259b-4e96-bb8c-3189735de365.png)

**Reply by conartist6** [reply]
Yep, that looks about right to me. I think that very technique will ultimately be used as part of this library, but I don't intend to replace the sync functions with it for the reasons I've mentioned.

**Reply by jcbhmr** [reply]
Sounds good! 👍
